File: backend/models/event.py
"""Universal normalized event model used by every connector."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime


@dataclass(slots=True)
class Event:
    title: str
    slug: str
    start_date: datetime
    timezone: str
    modality: str
    source_url: str
    source: str
    source_event_id: str
    event_hash: str
    description: str | None = None
    organization: str | None = None
    category: str = "Other"
    subcategory: str | None = None
    tags: list[str] = field(default_factory=list)
    event_type: str | None = None
    end_date: datetime | None = None
    venue: str | None = None
    city: str | None = None
    country: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    is_free: bool | None = None
    price: float | None = None
    currency: str | None = None
    registration_url: str | None = None
    image_url: str | None = None
    status: str = "published"

    def to_database_payload(self) -> dict[str, object]:
        """Return only columns accepted by the Phase 1 events table."""
        payload = asdict(self)
        payload["start_date"] = self.start_date.isoformat()
        if self.end_date is not None:
            payload["end_date"] = self.end_date.isoformat()
        return payload

File: frontend/src/types/event.ts
export type EventModality = 'in_person' | 'virtual' | 'hybrid'

export interface TechEvent {
  id: string
  title: string
  slug: string
  description: string | null
  organization: string | null
  category: string
  subcategory: string | null
  tags: string[]
  event_type: string | null
  start_date: string
  end_date: string | null
  timezone: string
  modality: EventModality
  venue: string | null
  city: string | null
  country: string | null
  is_free: boolean | null
  price: number | null
  currency: string | null
  registration_url: string | null
  source_url: string
  image_url: string | null
  source: string
  status: string
}

export type DateRange = 'all' | 'today' | 'week' | 'month'
export type SortOrder = 'ascending' | 'descending'
export type ViewMode = 'cards' | 'agenda'

export interface EventFilters {
  search: string
  category: string
  modality: EventModality | 'all'
  city: string
  organization: string
  eventType: string
  source: string
  dateRange: DateRange
  freeOnly: boolean
}

export const DEFAULT_FILTERS: EventFilters = {
  search: '',
  category: 'all',
  modality: 'all',
  city: 'all',
  organization: 'all',
  eventType: 'all',
  source: 'all',
  dateRange: 'all',
  freeOnly: false,
}

